### ISSUE TITLE
Linux support is not as broad as Godot's

### ISSUE THREAD

**Issue by mihe** [godot-jolt/godot-jolt#378]
This originated from the discussion in #371.

As of writing this I'm in the process of downgrading the CI builds from Ubuntu 22.04 to Ubuntu 20.04, in an effort to broaden Linux support to include Debian Stable. This is necessary because we're currently depending on glibc on Linux, and Ubuntu 22.04 includes glibc 2.35, whereas Ubuntu 20.04 and Debian Stable ("bullseye") include glibc 2.31, making them incompatible with builds from Ubuntu 22.04.

Godot itself uses [a custom SDK](https://github.com/godotengine/buildroot/) to build their stuff, which (at the time of writing) has support for Ubuntu versions as far back as Ubuntu 16.04.

To solve this problem indefinitely I'd like to explore something like creating a Docker container that contains a version of LLVM built with musl, statically link against musl, and then use libc++ instead of libstdc++. That should hopefully make the extension binaries completely self-contained and work on any version or distribution of Linux.

Whether this approach works with the various dependencies, or if the extension would even work in Godot at all, remains to be seen.

**Comment by Calinou** [comment]
> To solve this problem indefinitely I'd like to explore something like creating a Docker container that contains a version of LLVM built with musl, statically link against musl, and then use libc++ instead of libstdc++. That should hopefully make the extension binaries completely self-contained and work on any version or distribution of Linux.

Godot binaries link against glibc, so you don't need to go through the trouble of linking everything against musl.

**Comment by mihe** [comment]
I agree that the current level of support is probably Good Enough™, which is why I haven't rushed to fix this, but I'm still not thrilled about the idea of Linux support being bound to the runner images that GitHub Actions offers.

GitHub Actions only offer the two latest LTS versions of Ubuntu, so once Ubuntu 24.04 is released then I'll be forced to bump my Linux support as well, and I'm not sure how that will line up with Debian Stable and all that stuff. That'll also technically be a breaking change, which I'm not thrilled about.

**Comment by Calinou** [comment]
> GitHub Actions only offer the two latest LTS versions of Ubuntu, so once Ubuntu 24.04 is released then I'll be forced to bump my Linux support as well

I think you can run Docker containers within GitHub Actions, so you could run an old Ubuntu container within a 22.04 or even 24.04 image.

**Comment by mihe** [comment]
Yes, that was the plan with the musl stuff as well, but going "plain" 22.04 would be a reasonable compromise if the musl stuff doesn't work out I guess.

**Comment by DeeJayLSP** [comment]
I believe that at this point the custom SDK is usable with Godot Jolt. A new version got uploaded recently and building the plugin with it seems to go well.

**Comment by mihe** [comment]
Not being able to use the custom SDK was never the problem, as far as I know. The problem is that it uses GCC, which has worse codegen than Clang, with Clang getting around 15% higher numbers in Jolt's own `PerformanceTest` application.

In general I'd prefer using the same compiler on all platforms for the official builds, and I don't see this being enough of an issue in practice to sacrifice on performance for.

**Comment by DeeJayLSP** [comment]
I did a test by building the physics engine with the SDK (which has a version of libstdc++) on PATH, my own system's clang and `LIBCXX_FORCE_STATIC_LINK=1`. When testing and comparing to a normal build, the tests didn't show a significant difference (some slightly higher, some slightly slower).


Out of curiosity, does this make any difference when it comes to support (to better or worse)?

**Comment by mihe** [comment]
> When running `PerformanceTests` and comparing to the results of a normal build, the tests didn't show a significant difference (some slightly higher, some slightly slower).

I'm still seeing ~15% better results with Clang compared to the custom SDK (or the system GCC) on my (admittedly aging) Haswell-based Linux machine when comparing the `Discrete` multi-core results, which is the more relevant part of the test.

> Does this make any difference when it comes to support (to better or worse)?

I'm not quite sure what you're asking exactly, but using the custom SDK would technically speaking increase the range of support, yes. This new version of the SDK does however seem to bump glibc from version 2.19, which I believe should be compatible with Ubuntu as far back as 14.04, to now instead use version 2.28, which "only" ranges back to Ubuntu 18.10.

So by not using this new custom SDK this extension essentially removes support for Ubuntu versions 18.10, 19.04, 19.10, and of course equivalents in other distros, like Debian Buster (`oldoldstable`).

I don't personally see this being much of a problem. All of those Ubuntu versions are already end-of-life at this point, and in either case you'd likely be catering to a fraction of a fraction of a percentage of your potential userbase.

I can't find any official information of how SteamOS (Steam Deck etc.) deals with glibc though, but seeing as how it's Arch-based I assume they've long since upgraded to something newer than the glibc 2.31 that this extension targets.

**Comment by DeeJayLSP** [comment]
> I'm still seeing ~15% better results with Clang compared to the custom SDK

What I actually did was building two Clang versions, one being built with the SDK's libstdc++, not GCC.

**However**

At first I thought doing so would fix the compatibility since the resulting binary was different than the default one, until I tried to run inside Ubuntu containers from 16.04 up to 20.04, which failed to run on all of them.

Since the SDK is built with Buildroot, which heavily enforces GCC, it seems there is no way to use clang as a compiler with it (only as a library) due to [lack of headers](https://github.com/buildroot/buildroot/blob/cf635737ede4b1db7bdcb396a49f0abd0b62cd66/package/llvm-project/clang/clang.mk#L62).

Guess the SDK is out of consideration unless it gets clang someday.

**Comment by mihe** [comment]
I took a crack at making my own LLVM toolchain based on musl, as mentioned in the original issue description, and it (perhaps unsurprisingly) turned out to be more difficult than I thought it would be.

I was able (after hours of trial-and-error) to successfully build the entire toolchain and get something that actually generated a working and completely self-contained musl-based executable, but as soon as I try to throw anything moderately complex at it (like this project) I end up running into more problems.

With that I think I'll just fully commit to this discrepancy and mark this issue as `wontfix`. I feel much less conflicted about this now that Godot's custom toolchain is moving closer with its requirement of glibc 2.28.

If anyone desperately needs this they can always [build the extension](https://github.com/godot-jolt/godot-jolt/blob/master/docs/building.md#building-for-linux) themselves using GCC and Godot's toolchain.

I've also made the exact glibc version more explicit in `README.md`:

https://github.com/godot-jolt/godot-jolt/blob/f558d25fd21254e968d31136391ede2211305733/README.md#L84-L85
